Spigot-Server-Patches/0424-Seed-based-feature-search.patch

```diff
@@ -0,0 +1,113 @@
+diff --git a/src/main/java/net/minecraft/server/BiomeManager.java b/src/main/java/net/minecraft/server/BiomeManager.java
+index e96f544f1..68423645d 100644
+--- a/src/main/java/net/minecraft/server/BiomeManager.java
++++ b/src/main/java/net/minecraft/server/BiomeManager.java
+@@ -12,10 +12,12 @@ public class BiomeManager {
+         this.c = genlayerzoomer;
+     }
+ 
++    public BiomeManager withProvider(WorldChunkManager worldchunkmanager) { return a(worldchunkmanager); } // Paper - OBFHELPER
+     public BiomeManager a(WorldChunkManager worldchunkmanager) {
+         return new BiomeManager(worldchunkmanager, this.b, this.c);
+     }
+ 
++    public BiomeBase getBiome(BlockPosition blockposition) { return a(blockposition); } // Paper - OBFHELPER
+     public BiomeBase a(BlockPosition blockposition) {
+         return this.c.a(this.b, blockposition.getX(), blockposition.getY(), blockposition.getZ(), this.a);
+     }
+diff --git a/src/main/java/net/minecraft/server/ChunkCoordIntPair.java b/src/main/java/net/minecraft/server/ChunkCoordIntPair.java
+index f2a19acd8..09f1308b0 100644
+--- a/src/main/java/net/minecraft/server/ChunkCoordIntPair.java
++++ b/src/main/java/net/minecraft/server/ChunkCoordIntPair.java
+@@ -64,10 +64,12 @@ public class ChunkCoordIntPair {
+         }
+     }
+ 
++    public int getBlockX() { return d(); } // Paper - OBFHELPER
+     public int d() {
+         return this.x << 4;
+     }
+ 
++    public int getBlockZ() { return e(); } // Paper - OBFHELPER
+     public int e() {
+         return this.z << 4;
+     }
+diff --git a/src/main/java/net/minecraft/server/StructureGenerator.java b/src/main/java/net/minecraft/server/StructureGenerator.java
+index e8ce2ecf2..acfe732af 100644
+--- a/src/main/java/net/minecraft/server/StructureGenerator.java
++++ b/src/main/java/net/minecraft/server/StructureGenerator.java
+@@ -109,6 +109,15 @@ public abstract class StructureGenerator<C extends WorldGenFeatureConfiguration>
+                             if (flag1 || flag2) {
+                                 ChunkCoordIntPair chunkcoordintpair = this.a(chunkgenerator, seededrandom, j, k, i1, j1);
+                                 if (!world.getWorldBorder().isChunkInBounds(chunkcoordintpair.x, chunkcoordintpair.z)) { continue; } // Paper
++                                // Paper start - seed based feature search
++                                if (world.paperConfig.seedBasedFeatureSearch) {
++                                    BiomeManager biomeManager = world.getBiomeManager().withProvider(chunkgenerator.getWorldChunkManager());
++                                    BiomeBase biomeBase = biomeManager.getBiome(new BlockPosition(chunkcoordintpair.getBlockX() + 9, 0, chunkcoordintpair.getBlockZ() + 9));
++                                    if (!shouldGenerate(biomeManager, chunkgenerator, seededrandom, chunkcoordintpair.x, chunkcoordintpair.z, biomeBase)) {
```

Comment:
We need to synchronize on the StructureGenerator if we're generating a Stronghold due to some concurrency concerns here, but otherwise, this looks fine following the existing server logic here

Reply:
Aren't Strongholds handled separately (afaik it overrides the getNearestGeneratedFeature method due to it reading directly from the stronghold data file instead of searching chunks) or do you mean because of that?

Reply:
the shouldGenerate method ends up iterating the list of stronghold structures, actually looking at it, we already actually syncronize closer to the data modification itself, so shouldn't actually be needed

Reply:
Well it wouldn't ever be a problem in the first place as `WorldGenStronghold`'s `shouldGenerate` method isn't called by the general `getNearestGeneratedFeature` method ever as the `getNearestGeneratedFeature` is overriden by `WorldGenStronghold` with a different logic that isn't modified by my patch at all.